### PR TITLE
Authenticate users via OAuth2

### LINKED ISSUE #50: Authenticate administrators via OAuth
It'd make the {de,}provisioning process a lot easier if administrators could sign in via [OAuth](https://www.mediawiki.org/wiki/Help:OAuth) ([developer documentation](https://www.mediawiki.org/wiki/OAuth/For_Developers)). 

There're reference implementations for PHP, including examples that securely retrieve the user's group. This would also have the advantage of automatically revoking UTRS access to former administrators.

### PR DESCRIPTION
**Pull request by lfaraone**
This implements OAuth2 sign-in for UTRS, replacing the existing system. Fixes #50.

User accounts are automatically provisioned for members of the `sysop` group when they sign in. CheckUser status and email addresses are automatically propagated.

To use:
1. [Register an OAuth consumer](https://meta.wikimedia.org/wiki/Special:OAuthConsumerRegistration/propose) specifying:
   - Grants: "Authentication only with access to real name and email address via Special:OAuth/identify, no API access."
   - OAuth "callback" URL: `http://your-webserver/login.php`
   - Applicable project: `en.wikipedia.org`
2. Put your consumer token, secret token, and user agent into `public_html/src/config.js.php`
3. Visit `/login.php` in your favourite web browser.

Things this doesn't do:
- Remove the UI around password and email address configuration.
- Rip out the backend for logging in with a username and password; I don't want to disrupt other developers' workflows, so deferring to y'all whether that should be kept around.

Things I tested:
- [x] Accounts are auto-created, approved, activated on sign-in. 
- [x] Login does not work if the user is not a member of the required group (I changed `sysop` to `staff` below, for testing)
- [x] Already existing users' have their email address and CheckUser status updated on login.

Things I cannot test:
- Accepting the ToS. The relevant file doesn't appear to be part of the git repo; in any case, the wording will probably have to be tweaked. For testing, I manually frobbed the database after logging in.
- Actually using the app; even before I started working on OAuth2, logging in showed me [a fairly-broken display](https://cloud.githubusercontent.com/assets/73410/11328063/7e96c44c-917a-11e5-9025-0185f8dbdbd8.png)

### CODE REVIEW

File: public_html/login.php [hunk 1]
```diff
@@ -29,6 +30,135 @@
...
+            $user = User::getUserById($data['userID']);
+            if ($user->isCheckuser() !== $is_check || $user->getEmail() !== $payload->email) {
+                // XXX: Logging?
+                $query = $db->prepare("
+                        UPDATE user
+                        SET checkuser = :checkuser,
+                            email = :email
+                        WHERE userID = :userID");
+                $result = $query->execute(array(
+                        ':checkuser' => (bool)$is_check,
```
Comment (dqwiki) [inline public_html/login.php]: when you look at the DB, this is actually a number. Have we verified this line will consistently provide the right data type for the DB?

File: public_html/login.php [hunk 2]
```diff
@@ -29,6 +30,135 @@
...
+                    throw new UTRSDatabaseException($error);
+                }
+            }
+        }
+        header("Location: " . "home.php");
+        exit;
+    } else if ($is_admin === TRUE) {
+        $errors = 'You need to have a confirmed email address set in MediaWiki to use UTRS.';
+    } else {
+        $errors = 'Only administrators can review requests on UTRS.';
```
Comment (dqwiki) [inline public_html/login.php]: Watch this case happen, someone needs access for security/emergency reasons from the WMF (or hell uses their WMF account to signup) and doesn't have sysop privileges. How is this remedied?

File: public_html/src/userObject.php [hunk 3]
```diff
@@ -56,13 +56,20 @@ public function __construct(array $vars, $fromDB){
 			$this->toolAdmin = 0;
 			$this->checkuser = 0;
 			$this->developer = 0;
-			$this->useSecure = isset($vars['useSecure']);
+			$this->useSecure = !isset($vars['useSecure']);
```
Comment (dqwiki) [inline public_html/src/userObject.php]: I'm not sure what consequences come from this change for users to check whether or not to use secure. Either way it needs to be backwards compatible.

File: public_html/src/userObject.php [hunk 4]
```diff
@@ -56,13 +56,20 @@ public function __construct(array $vars, $fromDB){
...
 			$this->acceptToS = 1;
 			$this->replyNotify = 1;
 			$this->passwordHash = hash("sha512", $vars['password']);
 			$this->closed = 0;
 			$this->diff = $vars['diff'];
-			
+                        if ($oauth !== NULL) {
+                            $this->approved = 1;
+                            $this->active = 1;
```
Comment (dqwiki) [inline public_html/src/userObject.php]: At this point they have been verified that they are an admin right?

File: public_html/src/userObject.php [hunk 5]
```diff
@@ -56,13 +56,20 @@ public function __construct(array $vars, $fromDB){
...
+                            $this->checkuser = $oauth["checkuser"];
+                            $this->acceptToS = 0;
+                            $this->useSecure = 1;
+                            $this->passwordHash = "xxx";
```
Comment (dqwiki) [inline public_html/src/userObject.php]: We should change this to something more indicative of OAUTH like ":OAUTH:"

**Comment by dqwiki** [comment]
:bangbang: Requester attention needed

File: public_html/login.php (hunk 1 above)
Comment (lfaraone) [inline public_html/login.php]: Booleans don't exist in MySQL. The best you can do is `tinyint(1)`. When you cast a boolean to a string, you get this:

```
php > $foo = (int) TRUE;
php > echo $foo;
1
php > $foo = (int) FALSE;
php > echo $foo;
0
```

I assume that `->execute` DTRTs with a `bool`, but can cast to `(int)` if you'd prefer.

File: public_html/login.php (hunk 2 above)
Comment (lfaraone) [inline public_html/login.php]: Want me to add a check for `staff`?

File: public_html/src/userObject.php (hunk 3 above)
Comment (lfaraone) [inline public_html/src/userObject.php]: We should really just nix `useSecure`; sites with sensitive data should not allow usage via HTTP in 2015 (almost 2016).

Can we just redirect HTTP to HTTPS and I can rip this out entirely? :)

File: public_html/src/userObject.php (hunk 4 above)
Comment (lfaraone) [inline public_html/src/userObject.php]: Yes. `$oauth` is only non-`NULL` when called from the OAuth callsite.

File: public_html/src/userObject.php (hunk 3 above)
Comment (dqwiki) [inline public_html/src/userObject.php]: I just thought about it, that if we rip it out completely, there should be no backwards compatibility issues. Lets do that with this pull request.

File: public_html/login.php (hunk 2 above)
Comment (dqwiki) [inline public_html/login.php]: Sounds good, as long as you can check that global group.

File: public_html/login.php (hunk 1 above)
Comment (dqwiki) [inline public_html/login.php]: Just to be safe and consistent please.

**Comment by dqwiki** [comment]
:exclamation: New comments

**Comment by dqwiki** [comment]
Moving to #113

**Pull request closed without merging**